ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/CommandsValidator.scala

```diff
@@ -213,6 +223,63 @@ final class CommandsValidator(ledgerId: LedgerId) {
     } yield Submitters(actAs, readAs)
   }
 
+  private def validateDisclosedContracts(
+      disclosedContracts: Seq[ProtoDisclosedContract]
+  )(implicit
+      contextualizedErrorLogger: ContextualizedErrorLogger
+  ): Either[StatusRuntimeException, Set[DisclosedContract]] =
+    disclosedContracts.foldLeft[Either[StatusRuntimeException, Set[DisclosedContract]]](
+      Right(Set.empty[DisclosedContract])
+    )((contractz, contract) => {
+      for {
+        validatedContracts <- contractz
+        validatedContract <- validateDisclosedContract(contract)
+      } yield validatedContracts + validatedContract
+    })
+
+  private def validateDisclosedContract(
+      disclosedContract: ProtoDisclosedContract
+  )(implicit
+      contextualizedErrorLogger: ContextualizedErrorLogger
+  ): Either[StatusRuntimeException, DisclosedContract] = {
+    for {
+      templateId <- requirePresence(disclosedContract.templateId, "template_id")
+      validatedTemplateId <- validateIdentifier(templateId)
+      contractId <- requireContractId(disclosedContract.contractId, "contract_id")
+      createArguments <- requirePresence(disclosedContract.arguments, "arguments")
+      recordId <- validateOptionalIdentifier(createArguments.recordId)
+      validatedRecordField <- validateRecordFields(createArguments.fields)
+      // TODO DPP-1026 use correct metadata once the transaction stream actually outputs them
+      /*
+      metadata <- requirePresence(disclosedContract.metadata, "metadata")
+      createdAt <- requirePresence(metadata.createdAt, "created_at")
+      validatedCreatedAt = TimestampConversion.toLf(
+        createdAt,
+        TimestampConversion.ConversionMode.Exact,
+      )
+      keyHash <- validateHash(metadata.contractKeyHash, "contract_key_hash")
+       */
+    } yield DisclosedContract(
+      contractId = contractId,
+      contract = VersionedContractInstance(
+        template = validatedTemplateId,
+        arg = Versioned(
+          version = TransactionVersion.StableVersions.max, // TODO DPP-1026 what version to pick?
+          unversioned = Lf.ValueRecord(recordId, validatedRecordField),
+        ),
+        agreementText = "", // TODO DPP-1026 is this needed for anything?
```

Comment:
@moritzkiefer-da Similarly here, can we just make up an agreement text, or is its value used for validation?

Reply:
this one wouldn’t even matter if the engine tried to use it since only create nodes store the agreement text but not fetch & exercise. Also tracked as part of #13848